### PR TITLE
Fix bug where DynamicPool was leaking blocks (v1.0.1)

### PR DESCRIPTION
**Pull request by davidbeckingsale**

### CODE REVIEW

File: src/umpire/strategy/DynamicPool.cpp [hunk 1]
```diff
@@ -180,28 +180,27 @@ void* DynamicPool::allocate(std::size_t bytes)
...
     m_actual_bytes += alloc_bytes;
     m_curr_bytes += rounded_bytes;
 
-    const int64_t left_bytes{static_cast<int64_t>(
-        m_min_alloc_bytes - rounded_bytes)};
+    const std::size_t left_bytes{alloc_bytes - rounded_bytes};
 
     // Add free
-    if (left_bytes > m_align_bytes)
+    if (left_bytes > 0)
```
Comment (davidbeckingsale) [inline src/umpire/strategy/DynamicPool.cpp]: This was the cause of the issue. The other changes make the code consistent.

File: src/umpire/strategy/DynamicPool.cpp [hunk 2]
```diff
@@ -42,8 +42,8 @@ DynamicPool::DynamicPool(const std::string& name,
   m_actual_bytes{initial_alloc_bytes},
```
Comment (jdahm) [inline src/umpire/strategy/DynamicPool.cpp]: This should also get rounded.

**Comment by davidbeckingsale** [comment]
@mcfadden8 - this should be ready to go

**Comment by mcfadden8** [comment]
@davidbeckingsale, Why are build tests failing?

File: RELEASE_NOTES.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+# v1.0.1
+
+- Fixed a bug in DynamicPool where memory could be leaked.
```
Comment (mcfadden8) [inline RELEASE_NOTES.md]: Could you please be more descriptive here? What conditions would cause the leak?  I also think that the Jira ticket should be referenced.

File: scripts/make_release_tarball.sh [hunk 4]
```diff
@@ -7,7 +7,7 @@
 ##############################################################################
 
 TAR_CMD=gtar
-VERSION=1.0.0
+VERSION=1.0.1
```
Comment (mcfadden8) [inline scripts/make_release_tarball.sh]: Where is our "magic" version symbol maintained?  I presume it does NOT need to be updated since this is simply a patch-level release.  Is that correct?

File: src/umpire/strategy/DynamicPool.cpp [hunk 5]
```diff
@@ -33,17 +33,17 @@ DynamicPool::DynamicPool(const std::string& name,
                          const int align_bytes) noexcept :
   AllocationStrategy(name, id),
   m_allocator{allocator.getAllocationStrategy()},
-  m_min_alloc_bytes{min_alloc_bytes},
+  m_min_alloc_bytes{round_up(min_alloc_bytes, align_bytes)},
```
Comment (mcfadden8) [inline src/umpire/strategy/DynamicPool.cpp]: Where do we confirm that `align_bytes` is valid?  What if I set `align_bytes=21`, will that work?

Comment (mcfadden8) [inline src/umpire/strategy/DynamicPool.cpp]: Where do we validate `min_alloc_bytes`?  Is `min_alloc_bytes=0` expected to work?

File: src/umpire/strategy/DynamicPool.cpp [hunk 6]
```diff
@@ -33,17 +33,17 @@ DynamicPool::DynamicPool(const std::string& name,
...
   m_align_bytes{align_bytes},
   m_coalesce_heuristic{coalesce_heuristic},
   m_used_map{},
   m_free_map{},
   m_curr_bytes{0},
-  m_actual_bytes{initial_alloc_bytes},
+  m_actual_bytes{round_up(initial_alloc_bytes, align_bytes)},
```
Comment (mcfadden8) [inline src/umpire/strategy/DynamicPool.cpp]: Where do we validate `initial_alloc_bytes`? Is `initial_alloc_bytes=0` expected to work?

File: src/umpire/strategy/DynamicPool.cpp [hunk 7]
```diff
@@ -33,17 +33,17 @@ DynamicPool::DynamicPool(const std::string& name,
...
   m_highwatermark{0}
 {
-  insertFree(m_allocator->allocate(initial_alloc_bytes),
-             initial_alloc_bytes, true, initial_alloc_bytes);
+  const std::size_t bytes{round_up(initial_alloc_bytes, align_bytes)};
```
Comment (mcfadden8) [inline src/umpire/strategy/DynamicPool.cpp]: We need tests for edge-conditions of allocation sizes so that byte counts of 1 through `align_bytes-1` are tried.  I'm not sure if we have tested how well `round_up` works.

File: src/umpire/strategy/DynamicPool.cpp [hunk 8]
```diff
@@ -33,17 +33,17 @@ DynamicPool::DynamicPool(const std::string& name,
...
+  insertFree(m_allocator->allocate(bytes), bytes, true, bytes);
```
Comment (mcfadden8) [inline src/umpire/strategy/DynamicPool.cpp]:
```suggestion
  insertFree(m_allocator->allocate(rounded_to_alignment_bytes), rounded_to_alignment_bytes, true, rounded_to_alignment_bytes);
```

It does not need to be `rounded_to_alignment_bytes`, but I think that `bytes` by itself is not descriptive enough.

It could be `rounded_bytes` like it is in other parts of the code.

File: src/umpire/strategy/DynamicPool.cpp [hunk 9]
```diff
@@ -197,11 +196,10 @@ void* DynamicPool::allocate(std::size_t bytes)
     m_actual_bytes += alloc_bytes;
     m_curr_bytes += rounded_bytes;
 
-    const int64_t left_bytes{static_cast<int64_t>(
-        m_min_alloc_bytes - rounded_bytes)};
+    const std::size_t left_bytes{alloc_bytes - rounded_bytes};
```
Comment (mcfadden8) [inline src/umpire/strategy/DynamicPool.cpp]: Without this change, aren't allocations greater than `m_min_alloc_bytes` problematic?  It seems that this would also cause problems.  Are we able to test this somehow?

File: src/umpire/strategy/DynamicPool.cpp (hunk 1 above)
Comment (mcfadden8) [reply src/umpire/strategy/DynamicPool.cpp]: yikes.  I'm not sure what the intent of the previous line was.

Comment (mcfadden8) [reply src/umpire/strategy/DynamicPool.cpp]: I still could see `m_min_alloc_bytes - rounded_bytes` being something huge if `rounded_bytes > m_min_alloc_bytes`.

File: tests/integration/strategy_tests.cpp [hunk 10]
```diff
@@ -389,11 +389,43 @@ TEST(ReleaseTest, Works)
   void* ptr_two = alloc.allocate(1024);
   alloc.deallocate(ptr_two);
 
+  ASSERT_EQ(alloc.getCurrentSize(), 62);
   EXPECT_NO_THROW(alloc.release());
 
+  ASSERT_LE(alloc.getActualSize(), 64);
+
```
Comment (mcfadden8) [inline tests/integration/strategy_tests.cpp]: Why isn't this failing?  Should `alloc.getActualSize() == (1024+64)`?

File: tests/integration/strategy_tests.cpp [hunk 11]
```diff
@@ -389,11 +389,43 @@ TEST(ReleaseTest, Works)
...
+  ASSERT_EQ(allocator.getCurrentSize(), 0);
+  ASSERT_GE(allocator.getActualSize(), 0);
+
+  allocator.release();
+
+  ASSERT_EQ(allocator.getCurrentSize(), 0);
+  ASSERT_EQ(allocator.getActualSize(), 0);
+}
+
+
```
Comment (mcfadden8) [inline tests/integration/strategy_tests.cpp]: As mentioned in the code above, I think we should add a `TEST(ReleaseTest, AlignmentTest)` that would set the alignment size and then check allocations from 1 to alignment_bytes and make sure the getCurrentSize()/getActualSize() return believable results.

File: RELEASE_NOTES.md (hunk 3 above)
Comment (davidbeckingsale) [reply RELEASE_NOTES.md]: Agree.

File: scripts/make_release_tarball.sh (hunk 4 above)
Comment (davidbeckingsale) [reply scripts/make_release_tarball.sh]: Not merged yet.

File: src/umpire/strategy/DynamicPool.cpp (hunk 5 above)
Comment (davidbeckingsale) [reply src/umpire/strategy/DynamicPool.cpp]: What does "valid" mean?

Comment (davidbeckingsale) [reply src/umpire/strategy/DynamicPool.cpp]: No. And nowhere. I thought you only wanted a 1 line PR?

File: src/umpire/strategy/DynamicPool.cpp (hunk 7 above)
Comment (davidbeckingsale) [reply src/umpire/strategy/DynamicPool.cpp]: Okay.

File: src/umpire/strategy/DynamicPool.cpp (hunk 8 above)
Comment (davidbeckingsale) [reply src/umpire/strategy/DynamicPool.cpp]: Okay.

File: src/umpire/strategy/DynamicPool.cpp (hunk 9 above)
Comment (davidbeckingsale) [reply src/umpire/strategy/DynamicPool.cpp]: No. Allocations greater than `m_min_alloc_bytes` are exactly sized, so left_bytes would be > 0. In this case, left_bytes will now be 0. In any case, neither of these values would trigger the old `left_bytes > m_align_bytes` condition (which is correct).

File: tests/integration/strategy_tests.cpp (hunk 10 above)
Comment (davidbeckingsale) [reply tests/integration/strategy_tests.cpp]: No, `release` is called, and ptr_two has been deallocated.

File: scripts/make_release_tarball.sh (hunk 4 above)
Comment (mcfadden8) [reply scripts/make_release_tarball.sh]: Oh right.

File: src/umpire/strategy/DynamicPool.cpp (hunk 6 above)
Comment (jdahm) [reply src/umpire/strategy/DynamicPool.cpp]: Yes, I think we should UMPIRE_ERROR() if someone passes 0 as min_alloc_size, or align_bytes.

File: src/umpire/strategy/DynamicPool.cpp (hunk 5 above)
Comment (mcfadden8) [reply src/umpire/strategy/DynamicPool.cpp]: We make the rules for that since it is our API.  I will write a Jira ticket on this.  I'm no positive, but it it would not surprise me if round_up assumed align_bytes at least needed to be a power of 2.

Comment (mcfadden8) [reply src/umpire/strategy/DynamicPool.cpp]: I just wanted to point it out here since I am now thinking about the API.  I wrote a Jira ticket and we can decide where to address this elsewhere.

Comment (mcfadden8) [reply src/umpire/strategy/DynamicPool.cpp]: I wrote a Jira ticket.  We can decide when/if/where to address this later.

File: src/umpire/strategy/DynamicPool.cpp (hunk 6 above)
Comment (mcfadden8) [reply src/umpire/strategy/DynamicPool.cpp]: I wrote a Jira ticket to decide what the contract needs to be for all of these parameters.  We can decide when/if/where/how to address these separately.

File: src/umpire/strategy/DynamicPool.cpp (hunk 9 above)
Comment (mcfadden8) [reply src/umpire/strategy/DynamicPool.cpp]: um..  okay?

So, this caused a problem because the (evidently small) leak would cause the containing block to become hostage and would not ever be coallesced or released.  Is that right?

Comment (davidbeckingsale) [reply src/umpire/strategy/DynamicPool.cpp]: Yes - then repeating this over many many many allocations caused the large amounts we were seeing.

File: src/umpire/strategy/DynamicPool.cpp (hunk 2 above)
Comment (mcfadden8) [reply src/umpire/strategy/DynamicPool.cpp]: @davidbeckingsale, were you going to round this too as suggested by @jdahm ?

File: src/umpire/strategy/DynamicPool.cpp [hunk 12]
```diff
@@ -180,28 +180,27 @@ void* DynamicPool::allocate(std::size_t bytes)
...
 
-    if (left_bytes > m_align_bytes) {
+    if (left_bytes > 0) {
       insertFree(static_cast<unsigned char*>(ptr) + rounded_bytes, left_bytes,
                  false, whole_bytes);
     }
   } else {
     // Allocate new block -- note this does not check whether alignment is met
     const std::size_t alloc_bytes{std::max(rounded_bytes, m_min_alloc_bytes)};
+    std::cout << "Allocating " << alloc_bytes << " from resource for " << rounded_bytes << std::endl;
```
Comment (mcfadden8) [inline src/umpire/strategy/DynamicPool.cpp]: I assume we want to remove this right?

File: tests/integration/strategy_tests.cpp (hunk 11 above)
Comment (mcfadden8) [reply tests/integration/strategy_tests.cpp]: Can you add this test?  I think that adding a test would be good for the patch release if it is not too difficult.

File: tests/integration/strategy_tests.cpp (hunk 10 above)
Comment (mcfadden8) [reply tests/integration/strategy_tests.cpp]: Oh duh.  I missed the release() call.  I suppose we could consider adding another ASSERT_EQ before the release for getActualSize()?

File: src/umpire/strategy/DynamicPool.cpp (hunk 12 above)
Comment (davidbeckingsale) [reply src/umpire/strategy/DynamicPool.cpp]: Done already.

Comment (mcfadden8) [reply src/umpire/strategy/DynamicPool.cpp]: sorry, I don't no why my browser sometimes keeps old crap around

File: tests/integration/strategy_tests.cpp (hunk 11 above)
Comment (davidbeckingsale) [reply tests/integration/strategy_tests.cpp]: As discussed, we will test this elsewhere.

File: RELEASE_NOTES.md (hunk 3 above)
Comment (davidbeckingsale) [reply RELEASE_NOTES.md]: I have updated the description.

**Comment by mcfadden8** [comment]
Thanks @davidbeckingsale, the changes look great!  I plan to test this a little more and will approve afterwards.
